nexus/reconfigurator/planning/src/blueprint_builder/builder/storage_editor.rs

```diff
@@ -0,0 +1,249 @@
+use omicron_common::disk::DatasetName;
+use omicron_uuid_kinds::PhysicalDiskUuid;
+use omicron_uuid_kinds::SledUuid;
+use std::collections::BTreeMap;
+
+#[derive(Debug)]
+pub(super) struct BlueprintStorageEditor {
+    disks: BlueprintDisksEditor,
+    datasets: BlueprintDatasetsEditor,
+}
+
+impl BlueprintStorageEditor {
+    pub fn new(
+        disks: BTreeMap<SledUuid, BlueprintPhysicalDisksConfig>,
+        datasets: BTreeMap<SledUuid, BlueprintDatasetsConfig>,
+    ) -> Self {
+        Self {
+            disks: BlueprintDisksEditor::new(disks),
+            datasets: BlueprintDatasetsEditor::new(datasets),
+        }
+    }
+
+    pub fn sled_storage_editor<'a>(
+        &'a mut self,
+        sled_id: SledUuid,
+        sled_resources: &SledResources,
+        rng: &'a mut PlannerRng,
+    ) -> Result<SledStorageEditor<'a>, BlueprintDatasetsEditError> {
+        let disks = self.disks.sled_disks_editor(sled_id);
+        let datasets =
+            self.datasets.sled_datasets_editor(sled_id, sled_resources, rng)?;
+        Ok(SledStorageEditor { disks, datasets })
+    }
+
+    #[cfg(test)]
+    pub fn current_sled_disks(
+        &self,
+        sled_id: &SledUuid,
+    ) -> Option<&BTreeMap<PhysicalDiskUuid, BlueprintPhysicalDiskConfig>> {
+        self.disks.current_sled_disks(sled_id)
+    }
+
+    pub fn into_builders(
+        self,
+    ) -> (BlueprintDisksBuilder, BlueprintDatasetsBuilder) {
+        (
+            BlueprintDisksBuilder(self.disks),
+            BlueprintDatasetsBuilder(self.datasets),
+        )
+    }
+}
+
+#[derive(Debug)]
+pub(super) struct BlueprintDisksBuilder(BlueprintDisksEditor);
+
+impl BlueprintDisksBuilder {
+    pub fn build(
+        self,
+        sled_ids: impl Iterator<Item = SledUuid>,
+    ) -> BTreeMap<SledUuid, BlueprintPhysicalDisksConfig> {
+        self.0.build(sled_ids)
+    }
+}
+
+#[derive(Debug)]
+pub(super) struct BlueprintDatasetsBuilder(BlueprintDatasetsEditor);
+
+impl BlueprintDatasetsBuilder {
+    pub fn build(
+        self,
+        sled_ids: impl Iterator<Item = SledUuid>,
+    ) -> BTreeMap<SledUuid, BlueprintDatasetsConfig> {
+        self.0.build(sled_ids)
+    }
+}
+
+#[derive(Debug)]
+pub(super) struct SledStorageEditor<'a> {
+    disks: SledDisksEditor<'a>,
+    datasets: SledDatasetsEditor<'a>,
+}
+
+impl SledStorageEditor<'_> {
+    pub fn disk_ids(&self) -> impl Iterator<Item = PhysicalDiskUuid> + '_ {
+        self.disks.disk_ids()
+    }
+
+    pub fn ensure_disk(&mut self, disk: BlueprintPhysicalDiskConfig) -> Ensure {
+        let zpool = ZpoolName::new_external(disk.pool_id);
+
+        let result = self.disks.ensure_disk(disk);
+
+        // We ignore the result of possibly adding or updating the disk's
+        // dataset. Do we care to log if they've changed even if the disk
+        // doesn't?
+        self.datasets.ensure_debug_dataset(zpool.clone());
+        self.datasets.ensure_zone_root_dataset(zpool);
+
+        result
+    }
```

Comment:
I think we do care, especially for the purposes of keeping blueprint diffs accurate.

Would it be unreasonable to return:

```rust
struct DiskEnsure {
  disk: Ensure,
  dataset: Ensure,
}
```

? To basically propagate information about "what disks changed" separately from "what datasets changed"?

(... this also makes me think that `Ensure` should be typed, like our UUIDs, but that's a whole 'nother can of worms)

Reply:
I'd prefer to get rid of `Ensure` altogether and instead return the `EditCounts` recently added in PR 2/5 when `build` or `finalize` is called.

Reply:
Totally agree that's a better name for it, but I still think there's the question of scope (e.g., did you edit a disk, dataset, zone, etc?)

Reply:
> Totally agree that's a better name for it, but I still think there's the question of scope (e.g., did you edit a disk, dataset, zone, etc?)

Agreed. I think returning multiple `EditCounts` would make sense.

Reply:
> > I think we do care, especially for the purposes of keeping blueprint diffs accurate.
> 
> Can you say more about this? I think the return values of these functions are only really used in logging and some tests (which assert on the specific counts); diffs already do their own independent analysis of changes.

Which is not to say I'm opposed! Returning a combined edit counts thingamajig seems good, but it will need to be from a `finalize(self) -> AllEditCounts` kind of method, I think. Will give that a shot.

Reply:
yeah, I suppose not for the diffs, but the point of passing back the `ensure` structs in the first place was to keep track of "what changed", as you say, for tests, comments in the blueprint, etc.

(I'll take a look at the `AllEditCounts` stuff, but it sounds like we're on the same page here)